impeller/renderer/backend/vulkan/gpu_tracer_vk.cc

```diff
@@ -5,127 +5,155 @@
-    : context_(context) {
-  auto strong_context = context_.lock();
-  if (!strong_context) {
-    return;
-  }
-  timestamp_period_ = strong_context->GetPhysicalDevice()
+GPUTracerVK::GPUTracerVK(const std::shared_ptr<DeviceHolder>& device_holder)
+    : device_holder_(device_holder) {
+  timestamp_period_ = device_holder_->GetPhysicalDevice()
                           .getProperties()
                           .limits.timestampPeriod;
   if (timestamp_period_ <= 0) {
     // The device does not support timestamp queries.
     return;
   }
-  vk::QueryPoolCreateInfo info;
-  info.queryCount = kPoolSize;
-  info.queryType = vk::QueryType::eTimestamp;
-
-  auto [status, pool] = strong_context->GetDevice().createQueryPoolUnique(info);
-  if (status != vk::Result::eSuccess) {
-    VALIDATION_LOG << "Failed to create query pool.";
-    return;
-  }
-  query_pool_ = std::move(pool);
   // Disable tracing in release mode.
 #ifdef IMPELLER_DEBUG
   valid_ = true;
 #endif
 }
 
-void GPUTracerVK::RecordStartFrameTime() {
+bool GPUTracerVK::IsValid() const {
+  return valid_;
+}
+
+void GPUTracerVK::MarkFrameStart() {
+  in_frame_ = true;
+}
+
+void GPUTracerVK::MarkFrameEnd() {
   if (!valid_) {
     return;
   }
-  auto strong_context = context_.lock();
-  if (!strong_context) {
-    return;
-  }
-  auto buffer = strong_context->CreateCommandBuffer();
-  auto vk_trace_cmd_buffer =
-      CommandBufferVK::Cast(*buffer).GetEncoder()->GetCommandBuffer();
-  // The two commands below are executed in order, such that writeTimeStamp is
-  // guaranteed to occur after resetQueryPool has finished. The validation
-  // layer seem particularly strict, and efforts to reset the entire pool
-  // were met with validation errors (though seemingly correct measurements).
-  // To work around this, the tracer only resets the query that will be
-  // used next.
-  vk_trace_cmd_buffer.resetQueryPool(query_pool_.get(), current_index_, 1);
-  vk_trace_cmd_buffer.writeTimestamp(vk::PipelineStageFlagBits::eBottomOfPipe,
-                                     query_pool_.get(), current_index_);
-
-  if (!buffer->SubmitCommands()) {
-    VALIDATION_LOG << "GPUTracerVK: Failed to record start time.";
-  }
+  Lock lock(trace_state_mutex_);
+  current_state_ = (current_state_ + 1) % 16;
+
+  auto& state = trace_states_[current_state_];
+  FML_DCHECK(state.pending_buffers == 0u);
 
-  // The logic in RecordEndFrameTime requires us to have recorded a pair of
-  // tracing events. If this method failed for any reason we need to be sure we
-  // don't attempt to record and read back a second value, or we will get values
-  // that span multiple frames.
-  started_frame_ = true;
+  state.pending_buffers = 0;
+  state.current_index = 0;
+  state.contains_failure = false;
+  in_frame_ = false;
 }
 
-void GPUTracerVK::RecordEndFrameTime() {
-  if (!valid_ || !started_frame_) {
+void GPUTracerVK::RecordCmdBufferStart(const vk::CommandBuffer& buffer) {
+  if (!valid_ || !in_frame_) {
```

Comment:
Is the `!in_frame_` guard related to multiple concurrent frames?

If so can we add that as a 1-line comment. If not should it be DCHECK'd instead?